### PR TITLE
handle ssl errors when processing buffered data

### LINKED ISSUE #944: client does not disconnect if ssl error happens in Socket_continueWrites
```
20200903 182059.000 (3012523088)  (1)> MQTTAsync_cycle:4266
20200903 182059.000 (3012523088)   (2)> Socket_getReadySocket:249
20200903 182059.000 Return code 2 from read select
20200903 182059.000 (3012523088)    (3)> Socket_continueWrites:988
20200903 182059.000 (3012523088)     (4)> Socket_continueWrite:873
20200903 182059.000 (3012523088)      (5)> SSLSocket_continueWrite:1058
20200903 182059.000 (3012523088)       (6)> SSLSocket_error:100
20200903 182059.000 SSLSocket error (5) in SSL_write for socket 14 rc -1 errno 104 Connection reset by peer

20200903 182059.000 (3012523088)       (6)< SSLSocket_error:120 (-3)
20200903 182059.000 (3012523088)      (5)< SSLSocket_continueWrite:1073 (-1)
20200903 182059.104 (3012523088)     (4)< Socket_continueWrite:939 (-1)
20200903 182059.104 Freeing 80 bytes in heap at file /home/projects/pxc4/oe/build/workspace/sources/paho-mqtt-c/src/LinkedList.c line 224, heap use now 446400 bytes

20200903 182059.104 Freeing 16 bytes in heap at file /home/projects/pxc4/oe/build/workspace/sources/paho-mqtt-c/src/LinkedList.c line 229, heap use now 446320 bytes

20200903 182059.104 Freeing 16 bytes in heap at file /home/projects/pxc4/oe/build/workspace/sources/paho-mqtt-c/src/LinkedList.c line 224, heap use now 446304 bytes

20200903 182059.104 Freeing 16 bytes in heap at file /home/projects/pxc4/oe/build/workspace/sources/paho-mqtt-c/src/LinkedList.c line 229, heap use now 446288 bytes

20200903 182059.104 (3012523088)     (4)> MQTTAsync_writeComplete:1577
20200903 182059.104 (3012523088)      (5)> MQTTProtocol_checkPendingWrites:1495
20200903 182059.104 (3012523088)       (6)> MQTTProtocol_removePublication:287
20200903 182059.104 Freeing 13840 bytes in heap at file /home/projects/pxc4/oe/build/workspace/sources/paho-mqtt-c/src/MQTTProtocolClient.c line 290, heap use now 446272 bytes

20200903 182059.104 Freeing 64 bytes in heap at file /home/projects/pxc4/oe/build/workspace/sources/paho-mqtt-c/src/MQTTProtocolClient.c line 291, heap use now 432432 bytes

20200903 182059.104 Freeing 32 bytes in heap at file /home/projects/pxc4/oe/build/workspace/sources/paho-mqtt-c/src/LinkedList.c line 224, heap use now 432368 bytes

20200903 182059.104 Freeing 16 bytes in heap at file /home/projects/pxc4/oe/build/workspace/sources/paho-mqtt-c/src/LinkedList.c line 229, heap use now 432336 bytes

20200903 182059.104 (3012523088)       (6)< MQTTProtocol_removePublication:294
20200903 182059.104 Freeing 16 bytes in heap at file /home/projects/pxc4/oe/build/workspace/sources/paho-mqtt-c/src/LinkedList.c line 224, heap use now 432320 bytes

20200903 182059.104 Freeing 16 bytes in heap at file /home/projects/pxc4/oe/build/workspace/sources/paho-mqtt-c/src/LinkedList.c line 229, heap use now 432304 bytes

20200903 182059.104 (3012523088)      (5)< MQTTProtocol_checkPendingWrites:1512
20200903 182059.104 writeComplete: Removing response for msgid 0
20200903 182059.104 Freeing 16 bytes in heap at file /home/projects/pxc4/oe/build/workspace/sources/paho-mqtt-c/src/LinkedList.c line 229, heap use now 432288 bytes

20200903 182059.104 (3012523088)      (5)> MQTTProperties_free:382
20200903 182059.104 (3012523088)      (5)< MQTTProperties_free:405
20200903 182059.107 Freeing 96 bytes in heap at file /home/projects/pxc4/oe/build/workspace/sources/paho-mqtt-c/src/MQTTAsync.c line 1569, heap use now 432272 bytes

20200903 182059.107 (3012523088)     (4)< MQTTAsync_writeComplete:1680
20200903 182059.107 (3012523088)    (3)< Socket_continueWrites:1012 (0)
20200903 182059.107 Return code 1 from write select
20200903 182059.107 (3012523088)    (3)> isReady:225
20200903 182059.107 (3012523088)    (3)< isReady:230 (1)
20200903 182059.107 (3012523088)   (2)< Socket_getReadySocket:321 (14)
20200903 182059.107 m->c->connect_state = 0
20200903 182059.107 (3012523088)   (2)> MQTTPacket_Factory:112
20200903 182059.107 (3012523088)    (3)> WebSocket_getch:593
20200903 182059.107 (3012523088)     (4)> SSLSocket_getch:818
20200903 182059.107 (3012523088)      (5)> SocketBuffer_getQueuedChar:223
20200903 182059.107 (3012523088)      (5)< SocketBuffer_getQueuedChar:242 (-22)
20200903 182059.107 received 0 256 buflen 5

```

As can be seen from the logs above an SSL_FATAL error happens in SSLSocket_continueWrite. In my understanding we should not use this socket anymore after such an error. However the function `Socket_continueWrites` returns with 0 (no error). Because of this the error is ignored and the socket is still used. In this state my client was not able to receive any more messages.

I think the correct behavior would be to disconnect the client and call the connection lost callback like it's done when a fatal ssl error happens during a publish.

I can reproduce this issue by limiting the upload speed of the network which leads to tcp buffers filling up and then killing the connection.

What is your opinion on this?

### PR DESCRIPTION
**Pull request by sandro97git**
Signed-off-by: Sandro Scherer <sand.scherer@gmail.com>

Correctly handle ssl errors that occur when buffered messages (because of full tcp send buffer) are processed. See #944

### CODE REVIEW

File: src/MQTTAsync.c [hunk 1]
```diff
@@ -2552,7 +2552,8 @@ static thread_return_type WINAPI MQTTAsync_receiveThread(void* n)
 		if (rc == SOCKET_ERROR)
 		{
 			Log(TRACE_MINIMUM, -1, "Error from MQTTAsync_cycle() - removing socket %d", sock);
-			nextOrClose(m, rc, "socket error");
+			//nextOrClose(m, rc, "socket error");
+			MQTTAsync_disconnect_internal(m, 0);
```
Comment (sandro97git) [inline src/MQTTAsync.c]: @icraggs I'm not sure what the nextOrClose method does. Can you please explain? My expectation is that after an ssl error the client gets disconnected. This is also how it works for ssl errors that occur during normal publish.
https://github.com/eclipse/paho.mqtt.c/blob/master/src/MQTTAsync.c#L1996

Comment (icraggs) [reply src/MQTTAsync.c]: nextOrClose is called when a connection fails, to take the next action.  Closing the current connection, and then reconnecting with a new connection if that is needed.  A new connection is needed if MQTTVERSION_DEFAULT is set as that reconnects with MQTT 3.1, or automatic reconnect is set - so we kick off the automatic reconnect processing.

**Comment by icraggs** [comment]
Have you been able to try this out?

File: src/MQTTAsync.c (hunk 1 above)
Comment (sandro97git) [reply src/MQTTAsync.c]: I saw you recently added disconnect_internal in 275843846da2f5a0e4e27586f1e6fbd2f9e2f67b

**Comment by sandro97git** [comment]
> Have you been able to try this out?

Yes I can reproduce the misbehavior reliably and my changes resolve the problem.

**Comment by icraggs** [comment]
Great, thanks!

**Pull request merged**